### ISSUE TITLE
Manual: give more hierarchical structure to the "Language extension" chapter

### ISSUE THREAD

**Issue by shindere** [ocaml/ocaml#11164]
Follow-up to https://github.com/ocaml/ocaml/pull/11093#issuecomment-1090312943

It might be a good idea to re-structure this chapter of the manual and see
whether we can find a more logical and less flat organization.

One other question could be whether the "extensions" documented in thsi
chapter are still to be considered as "extensions" or whether their
documentation should be melt into other parts of the manual.

**Comment by smuenzel** [comment]
I tried to make a summary of each section of the extension manual and what ocaml version introduced the feature:

- Recursive definitions of values -- 1.0
- Recursive modules -- 3.07
- Private types -- 3.07, 3.11, 3.09
- Locally abstract types -- 3.12, new syntax in 4.03
- First-class modules -- 3.12, new stuff in 4.00,4.02,4.05
- Recovering the type of a module -- 3.12
- Substituting inside a signature -- 3.12, generalized in 4.06
- Type-level module aliases -- 4.02
- Overriding in open statements -- 4.01
- Generalized algebraic datatypes -- 4.00, refutation cases in 4.03, named existentials 4.13.0
- Syntax for Bigarray access -- 3.00
- Attributes -- 4.02, generalized notation in 4.03, some built-ins may have been introduced in other versions
- Extension nodes -- 4.02, generalized notation in 4.03, new infix notation in 4.04
- Extensible variant types -- 4.02, private in 4.06
- Generative functors -- 4.02
- Extension-only syntax -- 4.02.2, extended in 4.03, unary extension operators in 4.12.0, literals in 4.03
- Inline records -- 4.03
- Documentation comments -- 4.03
- Extended indexing operators -- 4.06
- Empty variant types -- 4.07.0
- Alerts -- 4.08
- Generalized open statements -- 4.08
- Binding operators  -- 4.08

I'm guessing it should probably be uncontroversial to move the features introduced in major version 3 into the non-extension sections of the manual.

**Comment by xavierleroy** [comment]
I don't agree with your guess.  For example recursive modules still don't work very well and still need a complete redesign, while GADTs are pretty stable at this point.  That the former was introduced before the latter is irrelevant.

**Comment by johnwhitington** [comment]
Previous discussion: https://github.com/ocaml/ocaml/issues/6676
